### ISSUE TITLE
`dotnet list package --outdated` doesn't work with sources that need auth

### ISSUE THREAD

**Issue by rohit21agrawal** [NuGet/Home#7605]
if a project has a source that needs to be authenticated, dotnet list package --outdated fails with the following error:

```
error: Unable to load the service index for source https://msazure.pkgs.visualstudio.com/_packaging/**********.
error:   Response status code does not indicate success: 401 (Unauthorized).
```

**Comment by nkolev92** [comment]
P0 : The easy fix is just adding the following in the command. https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/DeleteCommand.cs#L73. 

P2: The long fix is to rework the XPLAT commands to derive from 1 base command, similarly nuget.exe does and have it configured so it's not missed again. 

The easy fix should be done ASAP.

**Comment by nkolev92** [comment]
There are multiple moving parts to fix this issue. 
- [x] NuGet side change
- [x]  CLI side change https://github.com/dotnet/cli/pull/10721
- [ ] docs

The P0 NuGet side change is in PR.

**Comment by PatoBeltran** [comment]
as part of the long fix you mention, should xplat cli also read settings from the nuget.config to get credentials from there?

**Comment by nkolev92** [comment]
@PatoBeltran 
I think that's already being done. 
Each command individually determines which sources to use, and to do that, it reads the nuget.config.

**Comment by PatoBeltran** [comment]
@nkolev92 does that mean that this issue only exists when the user doesn't have the proper credentials stored in the config and needs to enter them interactively?

**Comment by nkolev92** [comment]
This is for users need to use the cross platform credential plugin, which is really the recommended way because clear text creds in configs are what we are trying to do away with.

**Comment by rrelyea** [comment]
@nkolev92 is driving CLI PRs for these. Fix in 5.0 is done.

**Comment by rrelyea** [comment]
nuget bug - fixed.
cli-sdk insertion - not yet.
docs - not yet.

**Comment by nkolev92** [comment]
A PR is out for the CLI side changes

**Comment by nkolev92** [comment]
https://github.com/dotnet/cli/pull/10721#issuecomment-473372010

CLI folks want to take the CLI change in 3.0.1xx. 

fyi @rrelyea We can try to argue otherwise or listen for feedback.

**Comment by nkolev92** [comment]
The NuGet wiring of the command is in. 
So once a successful restore is run which authenticates then the users should be able to do dotnet list with no problem. 

The --interactive switch will only happen in 3.0.1xx on CLI side.

**Comment by nkolev92** [comment]
CLI PR has been retarget to 3.0.1xx in https://github.com/dotnet/cli/pull/10994.

**Comment by nkolev92** [comment]
The NuGet and CLI work is done. 

The interactive parsing will be added in 3.0.100. 

Until then, a user should run any other command (namely restore needs run anyways) that supports --interactive. 
After that the plugins themselves will cache the PAT so when NuGet requests it for outdated it should just work.

**Comment by sebandraos** [comment]
I appreciate that the `--interactive` flag has now been added to this command (and works fine on `3.0.100-preview7-012821`) but the non-interactive version of the command is still causing issues (see [artifacts-credprovider issue](https://github.com/microsoft/artifacts-credprovider/issues/69))

Sample output with a few redactions:

```
PS C:\***> dotnet restore "C:\***.csproj" --source "https://pkgs.dev.azure.com/***/nuget/v3/index.json"
  Restore completed in 563.53 ms for C:\***.csproj.

PS C:\***> dotnet list "C:\***.csproj"  package
Project '***' has the following package references
   [net461]:
   Top-level Package               Requested    Resolved
   > PropertyChanged.Fody          3.0.1        3.0.1
   > ...

   [netstandard2.0]:
   Top-level Package                  Requested    Resolved
   > NETStandard.Library        (A)   [2.0.3, )    2.0.3
   > PropertyChanged.Fody             3.0.1        3.0.1
   > ...

(A) : Auto-referenced package.

PS C:\***> dotnet list "C:\***.csproj"  package --outdated --source "https://pkgs.dev.azure.com/***/nuget/v3/index.json"

The following sources were used:
   https://pkgs.dev.azure.com/***/nuget/v3/index.json

warn : The plugin credential provider could not acquire credentials. Authentication may require manual action. Consider re-running the command with --interactive for `dotnet`, /p:NuGetInteractive="true" for MSBuild or removing the -NonInteractive switch for `NuGet`
error: Unable to load the service index for source https://pkgs.dev.azure.com/***/nuget/v3/index.json.
error:   Response status code does not indicate success: 401 (Unauthorized).
```

As above, there is no issue with `restore` from the auth'ed feed nor with `list` without `--outdated` but it appears the cached credentials aren't accessed, the auth isn't sent with the list request or something else entirely isn't quite working here. This maybe an issue with the credential provider plugin in which case the issue can stay over there but as this issue looked very similar I figured it might be related.

If any more info is required please don't hesitate to let me know

**Comment by Gabriel-Lacatus** [comment]
I am also seeing these 401s happen randomly within Azure DevOps build pipelines. Sometimes we get the 401 and the build fails, then after a few minutes we run the same pipeline, without any changes to anything and it will work. It looks absolutely random, the only thing that I can think of that changes is the agent that's being assigned to run the pipeline but even so I've seen the same agent fail and then succeed.

**Comment by kmadof** [comment]
I have the same issue on Azure Pipelines. `dotnet restore` works well with configuration in `nuegt.config` file where I have defined sources. However, for `dotnet list package --outdated` or `dotnet list package --deprecated` i have `401`. 

As a workaround I did this:
```yaml
- script: |
	  del nuget.config
	  nuget sources add -name "Source1" -source https://pkgs.dev.azure.com/...../nuget/v3/index.json -username anything -password %SYSTEM_ACCESSTOKEN%
	  nuget sources add -name "Source2" -source https://pkgs.dev.azure.com/......./nuget/v3/index.json -username anything -password %SYSTEM_ACCESSTOKEN%
	enabled: true
	env:
	  SYSTEM_ACCESSTOKEN: $(System.AccessToken)
	displayName: Add NuGet source
```

So removing `nuget.config` and adding sources explicitly works. But why `dotnet restore` works and `dotnet list` not?

**Comment by rrelyea** [comment]
@sebandraos, @kmadof , @Gabriel-Lacatus - thanks for your reports. Will reopen to get more eyes on it.

**Comment by kmadof** [comment]
I found another workaround which is:

```yaml
      - script: nuget help
        displayName: 'Call nuget help in order to avoid timeout on fetching credential from provider'
      
      - task: NuGetAuthenticate@0

      - task: DotNetCoreCLI@2
        displayName: 'List deprecated packages'
        continueOnError: true
        inputs:
          command: custom
          custom: list
          arguments: 'package --deprecated'
          projects: '**/*.csproj'

      - task: DotNetCoreCLI@2
        displayName: 'List outdated packages'
        continueOnError: true
        inputs:
          command: custom
          custom: list
          arguments: 'package --outdated'
          projects: '**/*.csproj'
```

What is strange here, is if I run this without `nuget help` I will get timeout on fetching credentials from provider on the first run of `dotnet list package` but on succeeding all is fine.

**Comment by christianfosli** [comment]
This is still broken in dotnet 5.

`dotnet list package --outdated` 401's.
`dotnet list package --outdated --interactive` keeps asking for log-in for what seems like an endless loop, issuing a bunch of PAT's.

`dotnet restore --interactive` and `dotnet restore` works fine.

**Comment by OskarKlintrot** [comment]
Still broken in dotnet 6:
```
warn : The plugin credential provider could not acquire credentials. Authentication may require manual action. Consider re-running the command with --interactive for `dotnet`, /p:NuGetInteractive="true" for MSBuild or removing the -NonInteractive switch for `NuGet`
error: Response status code does not indicate success: 401 (Unauthorized).
```

**Comment by IFYates** [comment]
Still getting this issue in Azure Pipelines with dotnet SDK 6.0.301, but @kmadof's script is a decent workaround.

**Comment by nkolev92** [comment]
Hey all, 

This is a duplicate of https://github.com/NuGet/Home/issues/12090 which was fixed in 6.4 of NuGet and 7.0.100 of the .NET SDK. 

I'll close this issue now.


If you're having problems with .NET 7.0.100 SDK or later, please create a new issue and we'll be happy to take a look.
